### PR TITLE
CRM-21661, Do not set custom field of type float to 0

### PR DESCRIPTION
**Pull request by pradpnayak**
----------------------------------------
* CRM-21661:
  https://issues.civicrm.org/jira/browse/CRM-21661

Overview
----------------------------------------
_A brief description of the pull request. Try to keep it non-technical._

Before
----------------------------------------
screen shot 2018-01-17 at 1 12 58 am

After
----------------------------------------

screen shot 2018-01-17 at 1 14 10 am

### CODE REVIEW

File: CRM/Core/BAO/CustomValueTable.php [hunk 1]
```diff
@@ -217,6 +217,9 @@ public static function create(&$customParams) {
               break;
 
             default:
+              if ($value == NULL) {
```
Comment (eileenmcnaughton) [inline CRM/Core/BAO/CustomValueTable.php]: Can you explain this line (or better yet add code comments)

**Comment by eileenmcnaughton** [comment]
@pradpnayak this looks OK and it has been UI tested on per the JIRA comment. I think the logic is that if there IS a default for the field then the DB will pick it up so it does not need to be set 

if (!$value) {
          $value = 0;
        }

Probably that goes for money too - actually.

So I think this is OK & I could probably let it through without a test. I did have one code question though

**Comment by tschuettler** [comment]
@eileenmcnaughton Indeed, this issue should be part of: https://issues.civicrm.org/jira/browse/CRM-21223

File: CRM/Core/BAO/CustomValueTable.php (hunk 1 above)
Comment (pradpnayak) [reply CRM/Core/BAO/CustomValueTable.php]: When you pass NULL value to custom field of type float then you get fatal error with 
<code>
Jan 16 15:08:52  [info] $Fatal Error Details = Array
(
    [message] => One of parameters  (value: ) is not of the type Float
    [code] => 
)


Jan 16 15:08:52  [info] $backTrace = #0 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/CRM/Core/Error.php(378): CRM_Core_Error::backtrace("backTrace", TRUE)
#1 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/CRM/Utils/Type.php(504): CRM_Core_Error::fatal("One of parameters  (value: ) is not of the type Float")
#2 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/CRM/Core/DAO.php(1424): CRM_Utils_Type::validate("", "Float")
#3 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/CRM/Core/DAO.php(1301): CRM_Core_DAO::composeQuery("UPDATE civicrm_value_food_preference_2  SET soup_selection_4 = %2,number_fiel...", (Array:3), TRUE)
#4 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/CRM/Core/BAO/CustomValueTable.php(255): CRM_Core_DAO::executeQuery("UPDATE civicrm_value_food_preference_2  SET soup_selection_4 = %2,number_fiel...", (Array:3))
#5 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/CRM/Core/BAO/CustomValueTable.php(363): CRM_Core_BAO_CustomValueTable::create((Array:1))
#6 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/CRM/Event/BAO/Participant.php(228): CRM_Core_BAO_CustomValueTable::store((Array:2), "civicrm_participant", "54")
#7 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/CRM/Event/Form/Participant.php(1322): CRM_Event_BAO_Participant::create((Array:26))
#8 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/CRM/Event/Form/Participant.php(934): CRM_Event_Form_Participant->submit((Array:18))
#9 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/CRM/Core/Form.php(447): CRM_Event_Form_Participant->postProcess()
#10 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/CRM/Core/QuickForm/Action/Upload.php(169): CRM_Core_Form->mainProcess()
#11 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/CRM/Core/QuickForm/Action/Upload.php(136): CRM_Core_QuickForm_Action_Upload->realPerform(Object(CRM_Event_Form_Participant), "upload")
#12 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/packages/HTML/QuickForm/Controller.php(203): CRM_Core_QuickForm_Action_Upload->perform(Object(CRM_Event_Form_Participant), "upload")
#13 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/packages/HTML/QuickForm/Page.php(103): HTML_QuickForm_Controller->handle(Object(CRM_Event_Form_Participant), "upload")
#14 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/CRM/Core/Controller.php(351): HTML_QuickForm_Page->handle("upload")
#15 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/CRM/Event/Page/Tab.php(118): CRM_Core_Controller->run()
#16 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/CRM/Event/Page/Tab.php(175): CRM_Event_Page_Tab->edit()
#17 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/CRM/Core/Invoke.php(309): CRM_Event_Page_Tab->run((Array:4), NULL)
#18 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/CRM/Core/Invoke.php(84): CRM_Core_Invoke::runItem((Array:12))
#19 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/CRM/Core/Invoke.php(52): CRM_Core_Invoke::_invoke((Array:4))
#20 /Users/kapil/Sites/drupal-7.56/sites/all/modules/civicrm/drupal/civicrm.module(448): CRM_Core_Invoke::invoke((Array:4))
#21 [internal function](): civicrm_invoke("contact", "view", "participant")
#22 /Users/kapil/Sites/drupal-7.56/includes/menu.inc(527): call_user_func_array("civicrm_invoke", (Array:3))
#23 /Users/kapil/Sites/drupal-7.56/index.php(21): menu_execute_active_handler()
#24 {main}
</code>

Therefore i checked If Value == NULL and explicitly assigned 'Timestamp' to$type to bypass the error. This has been followed elsewhere in code(I know its a bad idea :( ).

eg: https://github.com/pradpnayak/civicrm-core/blob/76487bb9de979deef99205295b5a19bba3613b7c/CRM/Core/BAO/CustomValueTable.php#L158

**Comment by eileenmcnaughton** [comment]
@tschuettler yes - it feels like that line should not be there for money either - perhaps we are not setting the DB default to zero for those fields when creating the table & we should be (since it's kinda implicit)

**Comment by pradpnayak** [comment]
@eileenmcnaughton , @tschuettler I have updated code for money format as well.

**Comment by eileenmcnaughton** [comment]
@pradpnayak I'm still unsure about the change in the second file.

Can we confirm if existing installs do correctly have DB defaults for empty - we might need to go through & set mysql field defaults?

**Comment by pradpnayak** [comment]
@eileenmcnaughton  i had [replied](https://github.com/civicrm/civicrm-core/pull/11532#discussion_r162137188) to you about why change in second file was required. 

New install doesn't have number or Money custom fields. So i feel we can ignore. 

I don't think we need an upgrade code as we are not sure how many such fields are actually set as default 0 and how many are user inputted.

**Comment by eileenmcnaughton** [comment]
This is one of those issues that has stagnated because I'm not really comfortable with it. I feel like the RIGHT fix is to change our code so that a default is required for custom fields of type money & float & to remove that handling (that would also involve a DB update).

I could probably handle a change that checks the field metadata for a default & only defaults to 0 if none exists - although I would think a unit test would be required for that. 

The timestamp hack seems to be only required because we are not doing one of the above & it just seems really awful

**Comment by eileenmcnaughton** [comment]
Let's close this & re-visit it when we have a more complete fix + unit tests. This is tracked from JIRA so it's not lost but it's not really mergeable without a test

**Pull request closed without merging**